### PR TITLE
Replaces: `spellfile.vim` -> `spellfile.nvim`

### PR DESCRIPTION
**Pull request by cuducos**
As discussed in #7189, this PR suggests adding the code from [my prototype `spellfile.nvim`](https://github.com/cuducos/spellfile.nvim) to handle the download of missing spell files with Lua.

The major advantage is dropping the dependency of `netrw` for downloading spell files.

This PR is very raw, mostly for asking for help. 

In tests, how do I “import” the new module?

When it was a plugin named `spellfile_nvim` it was  `require'spellfile_nvim` — now… I left `local spellfile = require('spellfile')` but I am sure it does not work.

### CODE REVIEW

File: runtime/autoload/spellfile.lua
Comment (zeertzjq) [inline runtime/autoload/spellfile.lua:1]: This should be moved to runtime/lua/spellfile.lua

File: test/old/testdir/test_spellfile.vim
Comment (zeertzjq) [inline test/old/testdir/test_spellfile.vim:1]: Why is this file removed?

File: test/unit/spellfile_spec.lua
Comment (zeertzjq) [inline test/unit/spellfile_spec.lua:1]: This should be moved to test/functional

File: runtime/plugin/spellfile.lua [hunk 1]
```diff
@@ -0,0 +1,10 @@
+require('spellfile').setup()
+
+vim.api.nvim_create_autocmd('SpellFileMissing', {
+  callback = function()
+    local lang = vim.api.nvim_exec2('echo &spelllang', { output = true })
+    require('spellfile').load_file(lang.output)
+  end,
+})
+
+return { answer = 42 }
```
Comment (zeertzjq) [inline runtime/plugin/spellfile.lua:10]: ?

File: runtime/plugin/spellfile.lua [hunk 2]
```diff
@@ -0,0 +1,10 @@
+require('spellfile').setup()
```
Comment (zeertzjq) [inline runtime/plugin/spellfile.lua:1]: Removed this. It doesn't do anything.

File: runtime/autoload/spellfile.lua [hunk 3]
```diff
@@ -0,0 +1,137 @@
+local M = {}
```
Comment (clason) [inline runtime/autoload/spellfile.lua:1]: No need to expose all these functions (and in fact the configuration).

Also, you may want to wait for #29327.

File: runtime/plugin/spellfile.lua [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+vim.api.nvim_create_autocmd('SpellFileMissing', {
+  callback = function()
+    local lang = vim.api.nvim_exec2('echo &spelllang', { output = true })
```
Comment (clason) [inline runtime/plugin/spellfile.lua:5]: `vim.bo.spelllang` (ideally using `args.bufnr` from the callback argument)

File: runtime/autoload/spellfile.lua [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+      cmd = string.format('curl -fLo %s %s', pth, url)
+    elseif vim.fn.executable('wget') == 1 then
+      cmd = string.format('wget -O %s %s', pth, url)
+    else
+      vim.notify('No curl or wget found. Please install one of them.')
+      return
+    end
+
+    vim.notify(string.format('\nDownloading %s...', name))
+    vim.fn.system(cmd)
```
Comment (clason) [inline runtime/autoload/spellfile.lua:117]: use `vim.system` instead (if you don't want to wait for `vim.download`). No need to support `wget`, and `curl` support needs a healthcheck (see the existing `ripgrep` one; make sure to print the full output of `curl -v`).

File: runtime/autoload/spellfile.lua [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  return options[choice]
+end
+
+M.exists = function(file_name)
+  local is_file = function(pth)
+    local stat = vim.loop.fs_stat(pth)
+    return stat ~= nil and stat.type ~= nil and stat.type == 'file'
+  end
+
+  for _, dir in pairs(M.directory_choices()) do
```
Comment (clason) [inline runtime/autoload/spellfile.lua:54]: prefer `vim.fs` functions for iterating

File: runtime/autoload/spellfile.lua [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+  if choice <= 0 or choice > #options then
+    return
+  end
+
+  return options[choice]
+end
+
+M.exists = function(file_name)
+  local is_file = function(pth)
+    local stat = vim.loop.fs_stat(pth)
```
Comment (clason) [inline runtime/autoload/spellfile.lua:50]:
```suggestion
    local stat = vim.uv.fs_stat(pth)
```

File: runtime/autoload/spellfile.lua
Comment (cuducos) [reply runtime/autoload/spellfile.lua:1]: Great, fixed in c3fa93625

File: test/old/testdir/test_spellfile.vim
Comment (cuducos) [reply test/old/testdir/test_spellfile.vim:1]: Oops, accidentally removed that. Rolled back in c3fa93625

File: test/unit/spellfile_spec.lua
Comment (cuducos) [reply test/unit/spellfile_spec.lua:1]: True! Done in c3fa93625

File: runtime/autoload/spellfile.lua (hunk 3 above)
Comment (cuducos) [reply runtime/autoload/spellfile.lua]: Ok, this sounds stupid, but beginners make stupid decisions and questions, so here we go : )

I exposed a lot of functions so I could write tests for them. In other words, I don't know how to write tests if I don't expose and `require` in the test file. Any ideas or references so I can get to a better design here?

File: runtime/plugin/spellfile.lua (hunk 1 above)
Comment (cuducos) [reply runtime/plugin/spellfile.lua]: Ooop, my bad. Cleaned up in e1add1394.

File: runtime/plugin/spellfile.lua (hunk 2 above)
Comment (cuducos) [reply runtime/plugin/spellfile.lua]: Done in e1add1394

File: runtime/plugin/spellfile.lua (hunk 4 above)
Comment (cuducos) [reply runtime/plugin/spellfile.lua]: Done in e1add1394

File: runtime/autoload/spellfile.lua (hunk 5 above)
Comment (cuducos) [reply runtime/autoload/spellfile.lua]: I think this PR is far from being merged. I am happy working with `curl` by now and refactor late once #29327 gets merged : )

File: runtime/autoload/spellfile.lua (hunk 6 above)
Comment (cuducos) [reply runtime/autoload/spellfile.lua]: I'm gonna study `vim.fs` to see how to refactor. Probably gonna wait until I have tests working in this branch.

**Comment by cuducos** [comment]
Thank you @zeertzjq and @clason for the initial feedback. I pushed some commits with changes based on your comments. My main question at this moment is how to run tests. As I said in the [opening message](https://github.com/neovim/neovim/pull/29552#issue-2389557995):

> In tests, how do I “import” the new module?

Also, Christian pointed out I am exporting too many things — and I did that so they are testable. Is that a desired approach?

**Comment by clason** [comment]
It is not. Ideally, this should be completely opaque. (The file you deleted is testing everything relevant; as long as this passes, you are good. Configuration is a non-goal here!)

**Comment by cuducos** [comment]
Ok, gonna refactor with that in mind — and learn a bit about `vim.fs` in the way : ) Thank you so much

File: test/old/testdir/test_spellfile.vim [hunk 8]
```diff
@@ -1083,4 +1083,4 @@ func Test_mkspell_empty_dic()
 endfunc
 
 
-" vim: shiftwidth=2 sts=2 expandtab
+" vim: shiftwidth=2 sts=2 expandtab
```
Comment (zeertzjq) [inline test/old/testdir/test_spellfile.vim:1086]:
```suggestion
" vim: shiftwidth=2 sts=2 expandtab
```

Comment (zeertzjq) [inline test/old/testdir/test_spellfile.vim:1086]: Missing trailing EOL

**Comment by zeertzjq** [comment]
> It is not. Ideally, this should be completely opaque. (The file you deleted is testing everything relevant; as long as this passes, you are good. Configuration is a non-goal here!)

What? The builtin `spellfile.vim` is not covered by any tests. All tests still pass even if it is deleted.

File: test/old/testdir/test_spellfile.vim (hunk 8 above)
Comment (cuducos) [reply test/old/testdir/test_spellfile.vim:1086]: Out of curiosity: shouldn't StyLua take care of that? Asking because maybe there is something wrong in my setup and it didn't…

Comment (clason) [reply test/old/testdir/test_spellfile.vim:1086]: This is a vimscript file, not a Lua file...

Comment (cuducos) [reply test/old/testdir/test_spellfile.vim:1086]: Oh, brain fart. My bad. Sorry hahaha…

Comment (cuducos) [reply test/old/testdir/test_spellfile.vim]: Fixed in 65934874a ✨
